internal/state/storage/storage.go

```diff
@@ -0,0 +1,414 @@
+		zap.Int("endpoints", stats.NumEndpoints), zap.Duration("duration", duration))
+
+	storage.operationsMetric.WithLabelValues(operationSave, statusSuccess).Inc()
+	storage.lastSnapshotSizeMetric.WithLabelValues(operationSave).Set(float64(stats.Size))
+	storage.lastOperationClustersMetric.WithLabelValues(operationSave).Set(float64(stats.NumClusters))
+	storage.lastOperationAffiliatesMetric.WithLabelValues(operationSave).Set(float64(stats.NumAffiliates))
+	storage.lastOperationEndpointsMetric.WithLabelValues(operationSave).Set(float64(stats.NumEndpoints))
+	storage.lastOperationDurationMetric.WithLabelValues(operationSave).Set(duration.Seconds())
+
+	return nil
+}
+
+// Load loads all clusters' states from the persistent storage.
+func (storage *Storage) Load() (err error) {
+	defer func() {
+		if err != nil {
+			storage.operationsMetric.WithLabelValues(operationLoad, statusError).Inc()
+		}
+	}()
+
+	start := time.Now()
+
+	// open file for reading
+	file, err := os.Open(storage.path)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+
+	defer file.Close() //nolint:errcheck
+
+	stats, err := storage.Import(file)
+	if err != nil {
+		return fmt.Errorf("failed to read snapshot: %w", err)
+	}
+
+	if err = file.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+
+	duration := time.Since(start)
+
+	storage.logger.Info("state loaded", zap.Int("clusters", stats.NumClusters), zap.Int("affiliates", stats.NumAffiliates),
+		zap.Int("endpoints", stats.NumEndpoints), zap.Duration("duration", duration))
+
+	storage.operationsMetric.WithLabelValues(operationLoad, statusSuccess).Inc()
+	storage.lastSnapshotSizeMetric.WithLabelValues(operationLoad).Set(float64(stats.Size))
+	storage.lastOperationClustersMetric.WithLabelValues(operationLoad).Set(float64(stats.NumClusters))
+	storage.lastOperationAffiliatesMetric.WithLabelValues(operationLoad).Set(float64(stats.NumAffiliates))
+	storage.lastOperationEndpointsMetric.WithLabelValues(operationLoad).Set(float64(stats.NumEndpoints))
+	storage.lastOperationDurationMetric.WithLabelValues(operationLoad).Set(duration.Seconds())
+
+	return nil
+}
+
+// Import imports all clusters' states from the given reader.
+//
+// When importing, we avoid unmarshalling to the storagepb.StateSnapshot type directly, as it causes an allocation of all the cluster snapshots at once.
+// Instead, we process clusters in a streaming manner, unmarshaling them one by one and importing them into the state.
+func (storage *Storage) Import(reader io.Reader) (SnapshotStats, error) {
+	const maxHeaderSize = 20
+
+	size := 0
+	numClusters := 0
+	numAffiliates := 0
+	numEndpoints := 0
+	buffer := make([]byte, 0, maxHeaderSize)
+	reader = bufio.NewReader(reader)
+
+	// unmarshal the clusters in a streaming manner and import them into the state
+	if err := storage.state.ImportClusterSnapshots(func() (*storagepb.ClusterSnapshot, bool, error) {
+		var initialNumRead, additionalNumRead int
```

Comment:
I guess this can be simplified more, probably we can do a quick hack session?

Reply:
Updated one more time after our discussion.